### PR TITLE
Guided Tours: fix the `tourBranching` function to be able to descend render props

### PR DESCRIPTION
**Pull request by jsnajdr**
This patch fixes a regression in the `tourBranching` function that was introduced in #22862. That PR converted `Step` children to a functional render prop. That led to `tourBranching` being unable to descend into the children when examining the tour tree.

This gets fixed by `tourBranching` checking if `children` is a function, calling it when it is, and continuing to traverse the tree using the returned children.

Additional changes:
- rename the `config-parsing.js` file to `tour-branching.js` --- more appropriate name.
- simplify the algorithm how tour branches are collected and converted into objects.
- add a unit test for the `tourBranching` function!

**User-visible behavior**
The tour's ability to skip disabled or irrelevant steps was broken. Consider the following tour:
```jsx
<Tour name="jetpack">
  <Step name="init"><Next step="view-site" /></Step>
  <Step name="view-site" when={ isSitePreviewable }><Next step="themes" /></Step>
  <Step name="themes" />
</Tour>
```
When site is previewable, the `view-site` step is displayed. When it's not, clicking "Next" in `init` should skip all the way to `themes`. This was broken for 5 months since #22862. This patch makes that work again.

**How to test:**
Go to `/?tour=jetpack&flags=-preview-layout`. That should make `isSitePreviewable` return `false` and start the "Jetpack Basic" tour. Some steps should be skipped.

**Before the patch:** Instead of skipping the "View Site" step, the tour goes back to the initial step.
**After the patch:** The "View Site" step is skipped and the "Customize" item is higlighted and described.

### CODE REVIEW

File: client/layout/guided-tours/test/tour-branching.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+								<Continue step="categories-tags" hidden />
+							</ButtonRow>
+						</Fragment>
+					) }
+				</Step>
+
+				<Step name="categories-tags">
+					{ ( { translate } ) => (
+						<Fragment>
+							<p>{ translate( 'Categories and Tags' ) }</p>
```
Comment (a8ci18n) [inline client/layout/guided-tours/test/tour-branching.js]: Hi! I've found a possible matching string that has already been translated **92** times:
`translate( 'Categories and Tags Converter' )` *ES Score: 11*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/guided-tours-branching)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/layout/guided-tours/test/tour-branching.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+					{ ( { translate } ) => (
+						<Fragment>
+							<Continue step="finish">{ translate( 'Almost done' ) }</Continue>
+						</Fragment>
+					) }
+				</Step>
+				<Step name="finish">
+					{ ( { translate } ) => (
+						<Fragment>
+							<p>{ translate( 'You published your first blog post.' ) }</p>
```
Comment (a8ci18n) [inline client/layout/guided-tours/test/tour-branching.js]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'You published your first blog post' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/layout/guided-tours/tour-branching.js [hunk 3]
```diff
@@ -58,22 +56,32 @@ const fromPairsSequence = flow(
...
+/*
+ * From a JSX tree of a Tour, find how `Next` and `Continue` components link from one
+ * `Step` to another. For example, for a `<Step name="init" />` that contains a
+ * `<Next step='upload-image' />` button, the returned branching object will have a record:
+ * {
+ *   'init': { 'next': 'upload-image' }
+ * }
+ * This data is used to skip steps (`Step.skipToNext`) and to figure out if a step is the
+ * last one in a tour (`isLastStep` in tour context).
+ */
```
Comment (ockham) [inline client/layout/guided-tours/tour-branching.js]: Thanks for the comment, very helpful :+1:

**Review by ockham — APPROVED** [review]
LGTM, and tests well! :shipit:

**Pull request merged**